Python client

jmfrye opened the pull request:
First pass at Python 2.7 implementation of TreodeDB Client Library

BrainFlame commented:
There should be no need for multiple .gitignore files.  .gitignore inside sub folders are to add types that you don't want to ignore in the global .gitignore. The global .gitignore seems to already have .pyc .

Python2.7 should come with functools, although in the README it is suggested to install it.
On the other hand, I received an ImportError: No module named transaction. It is not clear what to install to get 'transaction' module. I tried pip install transaction and got me a zope.transaction module that does not work as expected.

jmfrye commented:
BrainFlame, I removed the extra .gitignore files.  Thanks.

I removed functools from the README.  Thanks again.

The transaction module is part of the client library.  I think the issue is that you cannot run the tests directly from the tests directory because they cannot find the modules in the parent folder.  That's a Python issue I don't know how to get around other than just run

python -m unittest discover 

instead.  Any better ideas?

BrainFlame commented:
Found the issue. Please rename the file Transaction.py to transaction.py (all lowercase).

jmfrye commented:
Done.  I had just renamed it locally before.

Topher-the-Geek commented:
Please rebase and squash your changes from the latest master branch.

Add a copyright header to all the .py files. You can find the text in the .scala files; just convert it to Python's comment format.

Topher-the-Geek commented on client/python/tx_clock.py:
```diff
@@ -0,0 +1,52 @@
+        self._input_in_usecs = input_in_usecs
+
+        # Default to current time
+        if (time == None):
```
Convert the system time to a long in the constructor, and drop the epsilon compare below.

Topher-the-Geek commented on client/python/tx_clock.py:
```diff
@@ -0,0 +1,52 @@
+
+        # Convert time to microseconds if needed
+        self.u_factor = 6
+        if (self._input_in_usecs):
```
I don't understand what this is about.

jmfrye replied:
Python long values don't have a maximum; I think I need to assign a max value for TxClock as a result:

http://stackoverflow.com/questions/9860588/maximum-value-for-long-integer

Topher-the-Geek commented on client/python/transaction.py:
```diff
@@ -0,0 +1,81 @@
+
+class Transaction(object):
+    def __init__(self, cache, read_timestamp=None, max_age=None, no_cache=False):
+        if (type(cache) != ClientCache):
```
I don't think you want to enforce this. In fact, I think you specifically want to pass in a mock of ClientCache for testing transaction.

Topher-the-Geek commented on client/python/tests/test_transaction.py:
```diff
@@ -0,0 +1,123 @@
+        cache = ClientCache(server)
+
+        # Mock the request method
+        headers = {
```
Mocking HTTP is necessary for testing the ClientCache. I think it's a hard way to go about testing the transaction. Can you mock ClientCache instead?

Topher-the-Geek commented on client/python/tests/test_cache_map.py:
```diff
@@ -0,0 +1,95 @@
+
+import random
+
+TEST_CNT = 10000
```
This abbreviation is likely to be pronounced cunt. Let's spell out count. Thanks.

Topher-the-Geek commented on client/python/tests/test_client_cache.py:
```diff
@@ -0,0 +1,213 @@
+        print "PASSED!"
+
+    # Test that calling read generates the expected request with headers
+    def test_client_cache_read(self):
```
I'm a bit confused about what the different test..read functions are each trying to achieve. I'd recommend that test names describe the input and expected behavior. Something like `Class_Inputs_Behavior`.

http://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html


http://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html

Topher-the-Geek commented on client/python/cache_map.py:
```diff
@@ -0,0 +1,176 @@
+
+    def _lru_remove(self, entries): 
+        lru_list = self.lru_list
+        for node in lru_list:
```
This looks like it's O(N). Removing from a doubly-linked list should be O(1). LRU eviction should be O(1).
